File: SPHAmod.py
#The iteration order is E=>D=>C
import numpy as np
from math import cosh,sinh,cos,sin,log
def md(x, y, z):
    """Creating SPHA"""
    return lambda *args: np.array(args[:2**(x+y+z)])
def md1(x, y, z):
    """SPHA Multiplication"""
    def mul(a, b):
        n = x + y + z
        if n == 0 : return np.array([a[0]*b[0]])
        if n == 1:
            if x == 1 : return np.array([a[0]*b[0] - a[1]*b[1], a[0]*b[1] + a[1]*b[0]])
            elif y == 1 : return np.array([a[0]*b[0] + a[1]*b[1], a[0]*b[1] + a[1]*b[0]])
            else : return np.array([a[0]*b[0], a[0]*b[1] + a[1]*b[0]])
        else:
            a1, a2 = a[:2**(n-1)], a[2**(n-1):]
            b1, b2 = b[:2**(n-1)], b[2**(n-1):]
            if z > 0:
                c1 = md1(x, y, z-1)(a1, b1)
                c2 = md1(x, y, z-1)(a1, b2) + md1(x, y, z-1)(a2, b1)
            elif y > 0:
                c1 = md1(x, y-1, z)(a1, b1) + md1(x, y-1, z)(a2, b2)
                c2 = md1(x, y-1, z)(a1, b2) + md1(x, y-1, z)(a2, b1)
            else:
                c1 = md1(x-1, y, z)(a1, b1) - md1(x-1, y, z)(a2, b2)
                c2 = md1(x-1, y, z)(a1, b2) + md1(x-1, y, z)(a2, b1)
            return np.concatenate((c1, c2))
    return mul
def md2(x, y, z):
    """SPHA Division"""
    def div(a, b):
        n = x + y + z
        if n == 0 : return np.array([a[0]/b[0]])
        if n == 1:
            if x == 1:
                denominator = b[0]**2 + b[1]**2
                return np.array([(a[0]*b[0] + a[1]*b[1]) / denominator,(a[1]*b[0] - a[0]*b[1]) / denominator])
            elif y == 1:
                denominator = b[0]**2 - b[1]**2
                return np.array([(a[0]*b[0] - a[1]*b[1]) / denominator,(a[1]*b[0] - a[0]*b[1]) / denominator])
            else : return np.array([a[0]*b[0] / b[0]**2,(a[1]*b[0] - a[0]*b[1]) / b[0]**2])
        else:
            b1 = b[:2**(n-1)]
            b2 = b[2**(n-1):]
            numerator = md1(x, y, z)(a, np.concatenate((b1, -b2)))
            if z > 0:
                denominator = md1(x, y, z-1)(b1, b1)
                return np.concatenate((md2(x, y, z-1)(numerator[:2**(n-1)], denominator), md2(x, y, z-1)(numerator[2**(n-1):], denominator)))
            elif y > 0:
                denominator = md1(x, y-1, z)(b1, b1) - md1(x, y-1, z)(b2, b2)
                return np.concatenate((md2(x, y-1, z)(numerator[:2**(n-1)], denominator), md2(x, y-1, z)(numerator[2**(n-1):], denominator)))
            else:
                denominator = md1(x-1, y, z)(b1, b1) + md1(x-1, y, z)(b2, b2)
                return np.concatenate((md2(x-1, y, z)(numerator[:2**(n-1)], denominator), md2(x-1, y, z)(numerator[2**(n-1):], denominator)))
    return div
def ps(index, x, y, z):#informal function
    if index >= 2**(x+y): return 0
    binary = bin(index)[2:].zfill(x+y+z)[y+z:]
    ones_count = binary.count('1')
    return (ones_count % 2 == 0)*2-1
def mp(x, y, z):
    """SPHA Exponentiation"""
    def pow(a, b):
        n = x + y + z
        result = np.zeros(2**n)
        result[0] = a**b[0]
        for i in range(1, 2**n):
            term = np.zeros(2**n)
            if ps(i, x, y, z) > 0:
                term[0] = cosh(b[i] * log(abs(a)))
                term[i] = sinh(b[i] * log(abs(a)))
            elif ps(i, x, y, z) < 0:
                term[0] = cos(b[i] * log(abs(a)))
                term[i] = sin(b[i] * log(abs(a)))
            else:
                term[0] = 1
                if a!=0:
                    term[i] = b[i] * log(a)
            result = md1(x, y, z)(result, term)
        return result
    return pow

File: HTML/SPHAmod.mjs
function md(x, y, z) {
    return (...args) => args.slice(0, Math.pow(2, x + y + z));
}

function md1(x, y, z) {
    return function mul(a, b) {
        const n = x + y + z;
        if (n === 0) return [a[0] * b[0]];
        if (n === 1) {
            if (x === 1) return [a[0]*b[0] - a[1]*b[1], a[0]*b[1] + a[1]*b[0]];
            else if (y === 1) return [a[0]*b[0] + a[1]*b[1], a[0]*b[1] + a[1]*b[0]];
            else return [a[0]*b[0], a[0]*b[1] + a[1]*b[0]];
        }
        const halfLen = Math.pow(2, n - 1);
        const a1 = a.slice(0, halfLen), a2 = a.slice(halfLen);
        const b1 = b.slice(0, halfLen), b2 = b.slice(halfLen);
        let c1, c2;
        if (z > 0) {
            c1 = md1(x, y, z-1)(a1, b1);
            c2 = md1(x, y, z-1)(a1, b2).map((v, i) => v + md1(x, y, z-1)(a2, b1)[i]);
        } else if (y > 0) {
            c1 = md1(x, y-1, z)(a1, b1).map((v, i) => v + md1(x, y-1, z)(a2, b2)[i]);
            c2 = md1(x, y-1, z)(a1, b2).map((v, i) => v + md1(x, y-1, z)(a2, b1)[i]);
        } else {
            c1 = md1(x-1, y, z)(a1, b1).map((v, i) => v - md1(x-1, y, z)(a2, b2)[i]);
            c2 = md1(x-1, y, z)(a1, b2).map((v, i) => v + md1(x-1, y, z)(a2, b1)[i]);
        }
        return [...c1, ...c2];
    };
}

function md2(x, y, z) {
    return function div(a, b) {
        const n = x + y + z;
        if (n === 0) return [a[0] / b[0]];
        if (n === 1) {
            if (x === 1) {
                const denominator = b[0]**2 + b[1]**2;
                return [(a[0]*b[0] + a[1]*b[1]) / denominator, (a[1]*b[0] - a[0]*b[1]) / denominator];
            } else if (y === 1) {
                const denominator = b[0]**2 - b[1]**2;
                return [(a[0]*b[0] - a[1]*b[1]) / denominator, (a[1]*b[0] - a[0]*b[1]) / denominator];
            } else {
                return [a[0]*b[0] / b[0]**2, (a[1]*b[0] - a[0]*b[1]) / b[0]**2];
            }
        }
        const halfLen = Math.pow(2, n - 1);
        const b1 = b.slice(0, halfLen), b2 = b.slice(halfLen);
        const numerator = md1(x, y, z)(a, [...b1, ...b2.map(v => -v)]);
        let denominator;
        if (z > 0) {
            denominator = md1(x, y, z-1)(b1, b1);
            return [
                ...md2(x, y, z-1)(numerator.slice(0, halfLen), denominator),
                ...md2(x, y, z-1)(numerator.slice(halfLen), denominator)
            ];
        } else if (y > 0) {
            denominator = md1(x, y-1, z)(b1, b1).map((v, i) => v - md1(x, y-1, z)(b2, b2)[i]);
            return [
                ...md2(x, y-1, z)(numerator.slice(0, halfLen), denominator),
                ...md2(x, y-1, z)(numerator.slice(halfLen), denominator)
            ];
        } else {
            denominator = md1(x-1, y, z)(b1, b1).map((v, i) => v + md1(x-1, y, z)(b2, b2)[i]);
            return [
                ...md2(x-1, y, z)(numerator.slice(0, halfLen), denominator),
                ...md2(x-1, y, z)(numerator.slice(halfLen), denominator)
            ];
        }
    };
}

function ps(index, x, y, z) {
    if (index >= Math.pow(2, x + y)) return 0;
    const binary = index.toString(2).padStart(x + y + z, '0');
    const c_part = binary.slice(y + z);
    const ones_count = (c_part.match(/1/g) || []).length;
    return (ones_count % 2 === 0) ? 1 : -1;
}

function mp(x, y, z) {
    return function pow(a, b) {
        const n = x + y + z;
        let result = new Array(Math.pow(2, n)).fill(0);
        if (a !== 0) {
            const abs_a = Math.abs(a);
            const log_abs_a = Math.log(abs_a);
            if (a > 0) {
                result[0] = Math.pow(a, b[0]);
            } else {
                result[0] = Math.pow(-a, b[0]) * Math.cos(b[0] * Math.PI);
                result[Math.pow(2, n-1)] = Math.pow(-a, b[0]) * Math.sin(b[0] * Math.PI);
            }
            for (let i = 1; i < Math.pow(2, n); i++) {
                let term = new Array(Math.pow(2, n)).fill(0);
                const ps_i = ps(i, x, y, z);
                if (ps_i > 0) {
                    term[0] = Math.cosh(b[i] * log_abs_a);
                    term[i] = Math.sinh(b[i] * log_abs_a);
                } else if (ps_i < 0) {
                    term[0] = Math.cos(b[i] * log_abs_a);
                    term[i] = Math.sin(b[i] * log_abs_a);
                } else {
                    term[0] = 1;
                    term[i] = b[i] * log_abs_a;
                }
                result = md1(x, y, z)(result, term);
                if (a < 0) {
                    const rotationFactor = Math.exp(ps_i * Math.PI * b[i]);
                    result = result.map(v => v * rotationFactor);
                }
            }
        } else {
            if (b.every(v => v === 0)) {
                result[0] = 1;
            } else {
                result[0] = 0;
            }
        }
        return result;
    };
}
export { md, md1, md2, mp } ;
